Docs: Exapnd GraphQL documentation (#6232)
solo-io/gloo: docs/content/guides/graphql/graphql_ui.md

## List GraphQL APIs in the Gloo Edge UI

The Gloo Edge UI is served from the `gloo-fed-console` service on port 8090. For more information about how to use the UI, see the [UI documentation]({{< versioned_link_path fromRoot="/guides/gloo_federation/read_only_console/" >}}).

1. Open the Gloo Edge UI.
   * `glooctl`: For more information, see the [CLI documentation]({{< versioned_link_path fromRoot="/reference/cli/glooctl_dashboard/" >}}).
     ```shell
     glooctl dashboard
     ```
   * `kubectl`: Port-forward the `gloo-mesh-ui` service on 8090.
     ```shell
     kubectl port-forward svc/gloo-fed-console -n gloo-system 8090:8090
     ```
2. Open your browser and connect to [http://localhost:8090](http://localhost:8090).
3. Optional: If authentication is enabled, sign in.
4. In the navigation bar, click **APIs**. The GraphQL, REST, and gRPC APIs in your Gloo Edge environment are listed.
5. Under _API Type_, select the GraphQL filter.
6. Review the overview details for each API, such as the namespace it is deployed in, the number of resolvers defined in the API, and the current status of the API.

> 7. In the **Actions** column, you can optionally download the configuration files for the API<!--, or delete the API configuration from your environment-->.

Author: TODO need to verify w the final UI cut whether the delete button is removed (didnt see it in the demo UI today)

Author: yep the latter. Totally feel free to change if you like!

![GraphQL APIs details screenshot]({{% versioned_link_path fromRoot="/img/screenshots/graphql_details.png" %}})

## Test GraphQL API functionality

Explore the functionality of an API by sending sample queries.

1. From the **APIs** overview page, click the name of a GraphQL API schema.
2. Click the **Explore** tab.
3. In the GraphiQL panel, you can specify example requests to send to the GraphQL API. The GraphiQL interface includes autocomplete based on the fields defined in your API configuration. For example, you might select one of your defined queries, and the fields within the query that you want data for.
4. Click the play button, which sends the request, and returns the response in the middle panel.
5. You can also explore the documentation for the API by clicking the **< Docs** buton to expand the Documentation Explorer panel.

## Create a GraphQL API

Define a new GraphQL API by using the UI.

1. From the **APIs** overview page, click **Create API**.
   1. Enter a name for the API.

>    <!--2. Select an excecutable API, such as for REST or gRPC services, or a stitched API, [TODO]-->

Author: TODO need to verify w the final UI cut whether type field is removed (didnt see it in the demo UI today)

   1. In the **API Details** tab, expand a configuration _Schema_ section. For example, you might start with the section for the top-level query.
   2. In the **Resolver** column, click **Define Resolver**.
   3. For the Resolver Type, choose a REST or gRPC resolver. 
   4. For the Upstream, choose a service for the upstream reference. The drop-down list is populated by the upstream services that are currently defined in your Gloo Edge environment.
   5. For the Resolver Config, fill out values for the provided fields. Note that you might not require all provided fields, and you can validate the syntax at any time by clicking **Validate**. Additionally, you can choose an existing resolver configuration that you already created from the drop-down list to modify for this field. For more information about how to configure each type of resolver, see [Manual schema configuration]({{< versioned_link_path fromRoot="/guides/graphql/resolver_config/" >}}).
   6. Click **Submit**.
   7. You can optonally click the **View Raw Config** button to verify that the resolver was added to your configuration.
   8. Repeat these steps to define a resolver for each configuration field.

> <!--4. To apply changes to your API configuration, toggle **Schema Introspection**, and click **Update**.-->

Author: TODO need to verify w the final UI cut whether the introspection toggle is removed (didnt see it in the demo UI today)

Author: @kcbabo Could you confirm this? additionally can I ask how we should explain introspection to the users? How can they set this if the toggle is removed (I saw tha even w/o the toggle that UI still showed the API w/ an introspection: enabled label) and what purpose does it have?

The Gloo Edge UI is served from the `gloo-fed-console` service on port 8090. For more information about how to use the UI, see the [UI documentation]({{< versioned_link_path fromRoot="/guides/gloo_federation/read_only_console/" >}}).

1. Open the Gloo Edge UI.
   * `glooctl`: For more information, see the [CLI documentation]({{< versioned_link_path fromRoot="/reference/cli/glooctl_dashboard/" >}}).
     ```shell
     glooctl dashboard
     ```
   * `kubectl`: Port-forward the `gloo-mesh-ui` service on 8090.
     ```shell
     kubectl port-forward svc/gloo-fed-console -n gloo-system 8090:8090
     ```
2. Open your browser and connect to [http://localhost:8090](http://localhost:8090).
3. Optional: If authentication is enabled, sign in.
4. In the navigation bar, click **APIs**. The GraphQL API schemas in your Gloo Edge environment are listed.
5. Review the overview details for each API, such as the namespace it is deployed in, the number of resolvers defined in the API, and the current status of the API.
6. In the **Actions** column, you can optionally download the configuration files for the API, or delete the API configuration from your environment.

> ![GraphQL APIs overview screenshot]({{% versioned_link_path fromRoot="/img/screenshots/TODO.png" %}})

Author: Will need screenshots for each of these sections still.

weight: 20
description: Visualize your GraphQL API and services in the GraphQL UI.
---

Use the Gloo Edge UI to review the health and configuration of your GraphQL APIs, test out your GraphQL API functionality, and create new GraphQL APIs.

## List GraphQL APIs in the Gloo Edge UI

The Gloo Edge UI is served from the `gloo-fed-console` service on port 8090. For more information about how to use the UI, see the [UI documentation]({{< versioned_link_path fromRoot="/guides/gloo_federation/read_only_console/" >}}).

1. Open the Gloo Edge UI.
   * `glooctl`: For more information, see the [CLI documentation]({{< versioned_link_path fromRoot="/reference/cli/glooctl_dashboard/" >}}).
     ```shell
     glooctl dashboard
     ```
   * `kubectl`: Port-forward the `gloo-mesh-ui` service on 8090.
     ```shell
     kubectl port-forward svc/gloo-fed-console -n gloo-system 8090:8090
     ```
2. Open your browser and connect to [http://localhost:8090](http://localhost:8090).
3. Optional: If authentication is enabled, sign in.
4. In the navigation bar, click **APIs**. The GraphQL, REST, and gRPC APIs in your Gloo Edge environment are listed.
5. Under _API Type_, select the GraphQL filter.

> 6. Review the overview details for each API, such as the namespace it is deployed in, the number of resolvers defined in the API, and the current status of the API.

Author: Custom API resource! `GraphQLApi` resource.

7. In the **Actions** column, you can optionally download the configuration files for the API<!--, or delete the API configuration from your environment-->.

<figure><img src="{{% versioned_link_path fromRoot="/img/screenshots/graphql_ov.png" %}}">
<figcaption style="text-align:center;font-style:italic">Figure: GraphQL APIs overview</figcaption></figure>

## Review GraphQL API details

Review the details of a GraphQL API, including its configuration, the resolvers for each query, and more.

1. From the list of GraphQL APIs, click the name of a GraphQL API schema.
2. In the **API Details** tab, you can view the raw configuration files, or explore the details of the _Schema_ and _Upstreams_ for the API.

>    * Configuration files: Click the **View Raw Config** button to view the raw configuration in the UI, and the **<file-name>.yaml** button to download the configuration YAML file.

Reviewer: The part in `<>` won't show...I tested this change in local preview and it worked.
```suggestion
   * Configuration files: Click the **View Raw Config** button to view the raw configuration in the UI, and the **&lt;file-name&gt;.yaml** button to download the configuration YAML file.
```

>    * _Schema_: To review the defined fields and values, click to expand each section. For example, you can expand a "Query" section to review the field names that are defined in the GraphQL query, the type of data returned by each field, and the resolver that processes the request and returns the data. Additionally, you can filter the displayed fields by using the search bar.

Reviewer: ```suggestion
   * _Schema_: To review the defined fields and values, click to expand each section. For example, you can expand **Query** section to review the field names that are defined in the GraphQL query, the type of data returned by each field, and the resolver that processes the request and returns the data. Additionally, you can filter the displayed fields by using the search bar.
```

   * _Upstreams_: To review the upstream services that the GraphQL server exposes, click the name of one of the listed services. The **Upstreams** page for the service opens. For more information about the upstream services page, see [Exploring Virtual Services and Upstreams]({{< versioned_link_path fromRoot="/guides/gloo_federation/read_only_console/#exploring-virtual-services-and-upstreams" >}}).

<figure><img src="{{% versioned_link_path fromRoot="/img/screenshots/graphql_details.png" %}}">
<figcaption style="text-align:center;font-style:italic">Figure: GraphQL API details page</figcaption></figure>

## Test GraphQL API functionality

Explore the functionality of an API by sending sample queries.

1. From the **APIs** overview page, click the name of a GraphQL API schema.
2. Click the **Explore** tab.

> 3. In the GraphiQL panel, you can specify example requests to send to the GraphQL API. The GraphiQL interface includes autocomplete based on the fields defined in your API configuration. For example, you might select one of your defined queries, and the fields within the query that you want data for.

Reviewer: ```suggestion
3. In the GraphQL panel, you can specify example requests to send to the GraphQL API. The GraphQL interface autocompletes based on the fields defined in your API configuration. For example, you might select one of your defined queries, and the fields within the query that you want data for.
```

Author: So this is specifically called `GraphiQL` which is an [explorer integration](https://github.com/graphql/graphiql)

Reviewer: Ok, I might add this in the shortdesc for this section.

4. Click the play button, which sends the request, and returns the response in the middle panel.

> 5. You can also explore the documentation for the API by clicking the **< Docs** buton to expand the Documentation Explorer panel.

Reviewer: ```suggestion
5. You can also explore the documentation for the API by clicking the **< Docs** button to expand the Documentation Explorer panel.
```

## Create a GraphQL API

Define a new GraphQL API by using the UI.

1. From the **APIs** overview page, click **Create API**.

>    1. Enter a name for the API.
>    <!--2. Select an excecutable API, such as for REST or gRPC services, or a stitched API, [TODO]-->

Reviewer: As-is, this comment breaks the subbullet formatting.
```suggestion
   1. Enter a name for the API.<!--2. Select an excecutable API, such as for REST or gRPC services, or a stitched API, [TODO]-->
```

   3. Click **Upload Schema** to add a `.gql` configuration file.
2. Click **Create API**. The details page for the API opens.
3. If no resolvers are defined, you might see a warning. To define a resolver:
   1. In the **API Details** tab, expand a configuration _Schema_ section. For example, you might start with the section for the top-level query.
   2. In the **Resolver** column, click **Define Resolver**.
   3. For the Resolver Type, choose a REST or gRPC resolver. 
   4. For the Upstream, choose a service for the upstream reference. The drop-down list is populated by the upstream services that are currently defined in your Gloo Edge environment.

>    5. For the Resolver Config, fill out values for the provided fields. Note that you might not require all provided fields, and you can validate the syntax at any time by clicking **Validate**. Additionally, you can choose an existing resolver configuration that you already created from the drop-down list to modify for this field. For more information about how to configure each type of resolver, see [Manual schema configuration]({{< versioned_link_path fromRoot="/guides/graphql/resolver_config/" >}}).

Reviewer: This mentions a **Validate** button, but the screenshot shows this tab, and I don't see one.

Author: Hmmm. I think this may show up as you are editing the resolver config, but not when the config is already submitted... I'm honestly not sure and when I check in the demo UI just now, I wasnt able to edit _any_ of the configs (i imagine bc i dont have access to edit anything in that cluster).

@kcbabo Do you know if the validate button exists in the "final" cut of the UI?
